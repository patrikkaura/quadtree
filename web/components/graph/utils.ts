export function getRange(start: number, end: number, step: number) {
  return Array.from(
    { length: (end - start) / step + 1 },
    (_, i) => start + i * step
  );
}

export function getGraphScale(sequenceLength: number) {
  if (sequenceLength < 1000) {
    return 10;
  } else if (sequenceLength < 2500) {
    return 100;
  }
  return 250;
}
