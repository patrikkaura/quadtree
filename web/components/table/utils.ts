export const groupNucleicPairs = (sequence: string) => {
  const groups: string[][] = [];

  sequence.split("").reduce<string[]>((acc, char, index) => {
    if (!acc.length) {
      return [char];
    }  else if (acc[0] === char) {
      acc.push(char);
      return acc;
    } else {
      groups.push(acc);
      return [char];
    }
  }, []);

  return groups;
};
