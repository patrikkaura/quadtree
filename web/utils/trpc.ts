import { httpBatchLink } from "@trpc/client";
import { createTRPCNext } from "@trpc/next";

import type { AppRouter } from "../pages/api/trpc/[trpc]";

function getBaseUrl() {
  if (typeof window !== "undefined") {
    // browser should use relative path
    return "";
  }
  if (process.env.VERCEL_URL) {
    // reference for vercel.com
    return `https://${process.env.VERCEL_URL}`;
  }
  // assume localhost
  return `http://localhost:${process.env.PORT ?? 3000}`;
}

export const trpc = createTRPCNext<AppRouter>({
  config({ ctx }) {
    return {
      links: [
        httpBatchLink({
          url: `${getBaseUrl()}/api/trpc`,
        }),
      ],
    };
  },
});
