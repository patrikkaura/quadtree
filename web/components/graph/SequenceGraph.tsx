import {
  CategoryScale,
  Chart as ChartJS,
  Legend,
  LinearScale,
  LineElement,
  PointElement,
  Title,
  Tooltip,
} from "chart.js";
import { useMemo } from "react";
import { Line } from "react-chartjs-2";

import { getGraphScale, range } from "./utils";

ChartJS.register(
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend
);

type Props = {
  intervals: number[][];
  length: number;
};

export function SequenceGraph({ intervals, length }: Props) {
  console.log({ length });

  const graphScale = useMemo(() => getGraphScale(length), [length]);

  let labels = useMemo(() => {
    let flattenedIntervals = intervals
      .map(([start, end]) => [start - 1, start, end, end + 1])
      .flat(1);

    flattenedIntervals = [
      ...range(0, length, graphScale),
      ...flattenedIntervals,
    ].sort((a, b) => a - b);

    flattenedIntervals = [
      ...flattenedIntervals,
      ...intervals.map(([start, end]) => Math.floor((start + end) / 2)),
    ];

    return flattenedIntervals.sort((a, b) => a - b);
  }, [intervals, length, graphScale]);

  const detectedIntervals: Record<number, number> = intervals.reduce(
    (acc, [start, end]) => {
      return {
        ...acc,
        [Math.floor((start + end) / 2)]: end - start,
      };
    },
    {}
  );

  const data = {
    labels,
    datasets: [
      {
        label: "Putative G4",
        data: labels.map((position) => detectedIntervals[position] ?? 0),
        borderColor: "rgb(103, 134, 198)",
        backgroundColor: "rgba(255, 255, 255, 1)",
      },
    ],
  };

  return (
    <Line
      height={100}
      options={{
        responsive: true,
        plugins: {
          legend: {
            position: "top" as const,
          },
          title: {
            display: true,
            text: "Quadtree",
          },
        },
        normalized: true,
      }}
      data={data}
    />
  );
}
