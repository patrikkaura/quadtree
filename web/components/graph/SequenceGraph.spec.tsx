import { render } from "@testing-library/react";

import { Props, SequenceGraph } from "./SequenceGraph";
import { getGraphScale, getRange } from "./utils";

jest.mock("react-chartjs-2", () => ({
  Line: () => null,
}));

jest.mock("./utils");

const mockedGetGraphScale = jest
  .mocked(getGraphScale)
  .mockImplementation(() => 10);
const mockedGetRange = jest
  .mocked(getRange)
  .mockImplementation(() => [0, 1, 2, 3, 4, 5]);

describe("SequenceGraph", () => {
  let props: Props = {
    intervals: [
      [10, 20],
      [30, 40],
      [50, 60],
      [70, 80],
    ],
    length: 100,
  };

  it("should render", () => {
    const { container } = render(<SequenceGraph {...props} />);
    expect(container).toMatchSnapshot();

    expect(mockedGetRange).toHaveBeenCalledTimes(1);
    expect(mockedGetRange).toHaveBeenCalledWith(0, props.length, 10);

    expect(mockedGetGraphScale).toHaveBeenCalledWith(100);
    expect(mockedGetGraphScale).toHaveBeenCalledTimes(1);
  });
});
