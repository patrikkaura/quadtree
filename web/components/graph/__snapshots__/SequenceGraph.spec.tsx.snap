// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SequenceGraph should render 1`] = `
<div>
  <div
    class="border px-6 py-4 shadow-sm rounded-xl p-2 "
  />
</div>
`;
